### PR TITLE
Fix use of milestone step outside parallel [JENKINS-38464]

### PR DESCRIPTION
**Pull request by svanoort**
Solves [JENKINS-38464](https://issues.jenkins-ci.org/browse/JENKINS-38464)

Also takes up the new FlowScanner APIs for speed+correctness.

### CODE REVIEW

**Comment by svanoort** [comment]
@reviewbybees especially @amuniz and @jglick

File: pom.xml [hunk 1]
```diff
@@ -30,13 +30,13 @@
     <repositories>
         <repository>
             <id>repo.jenkins-ci.org</id>
-            <url>http://repo.jenkins-ci.org/public/</url>
+            <url>https://repo.jenkins-ci.org/public/</url>
```
Comment (jglick) [inline pom.xml]: I also recommend updating the parent POM version.

File: src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java [hunk 2]
```diff
@@ -85,30 +89,15 @@ public Void run() throws Exception {
...
-                parallelDetectionEnabled--;
-            }
-
-            OrdinalAction a = n.getAction(OrdinalAction.class);
-            if (a != null) {
-                previousOrdinal = a.ordinal;
-                break;
-            }
+        List<FlowNode> heads = node.getExecution().getCurrentHeads();
+        if (heads.size() > 1) {  // TA-DA!  We're inside a parallel, which is forbidden.
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java]: It is not immediately obvious to me that this is correct. What if the other branch has completed? I suppose the current heads would include this step plus the other branch’s block end node?

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java]: FWIW the check I envisioned here was something along the lines of

```java
FlowScanningUtils.fetchEnclosingBlocks(node).
    filter(FlowScanningUtils.hasActionPredicate(ThreadNameAction.class)).
    hasNext()
```

**Comment by amuniz** [comment]
LGTM 👍 and :bee: (after the tests pass)

File: src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java (hunk 2 above)
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepExecution.java]: > current heads would include this step plus the other branch’s block end node?

Correct.  See also https://github.com/jenkinsci/workflow-api-plugin/pull/20 for an example of how we work with this.

**Comment by svanoort** [comment]
Close and re-open simply because tests completed successfully... but were never reported back to the PR.

**Comment by svanoort** [comment]
@reviewbybees done

**Pull request merged**